internal/contour/virtualhost.go

```diff
@@ -119,7 +122,7 @@ func (v *VirtualHostCache) recomputevhost(vhost string, ingresses map[metadata]*
 
 // action computes the cluster route action, a *route.Route_route for the
 // supplied ingress and backend.
-func action(i *v1beta1.Ingress, be *v1beta1.IngressBackend) *route.Route_Route {
+func action(i *v1beta1.Ingress, be *v1beta1.IngressBackend, useWebsocket *types.BoolValue) *route.Route_Route {
```

Comment:
please make a wrapper here, say `websocketaction` that calls `action` then applies `ca.Route.UseWebsocket = true`. See `clusteraction` in virtualhost_test.go

Reply:
`clusteraction` is being used only in `virtualhost_test.go` and I already created a sibling for websocket tests called `clusteractionwebsocket` I can rename it to `websocketaction`:

```
// clusteractionwebsocket returns a cluster action with websocket support.
func clusteractionwebsocket(name string) *route.Route_Route {
	c := clusteraction(name)
	c.Route.UseWebsocket = &types.BoolValue{Value: true}
	return c
}
```

If I create a new func `websocketaction` the loop is going to look very awful with ifs here and there. Do you really prefer this change?

```
if wr[p.Path] {
    vv.Routes = append(vv.Routes, route.Route{
        Match:  pathToRouteMatch(p),
        Action: websocketaction(ing, &p.Backend),
    })
} else {
    vv.Routes = append(vv.Routes, route.Route{
        Match:  pathToRouteMatch(p),
        Action: action(ing, &p.Backend),
    })
}
```

```
if wr["/"] {
    r := route.Route{
        Match:  prefixmatch("/"),
        Action: websocketaction(i, i.Spec.Backend),
    }
} else {
    r := route.Route{
        Match:  prefixmatch("/"),
        Action: action(i, i.Spec.Backend, wr["/"]),
    }
}
```

```
if wr[p.Path] {
    r := route.Route{
        Match:  pathToRouteMatch(p),
        Action: websocketaction(i, &p.Backend),
    }
} else {
    r := route.Route{
        Match:  pathToRouteMatch(p),
        Action: action(i, &p.Backend),
    }
}
```

Doesn't look like a very DRY code.